src/compiler/transformers/declarations.ts

```diff
@@ -973,10 +973,124 @@ namespace ts {
                             input.isTypeOf
                         ));
                     }
+                    case SyntaxKind.UnionType: {
+                        if(input.types.length <=1) return cleanup(input);
+                        const clone = getMutableClone(input);
+                        const types = clone.types.map(n=>n as TypeReferenceNode);
+                        types.sort((a,b) => {
+                            // TODO Undefined values are last, add implementation to getSortableName to support them.
+                            const keyA = getSortableName(a);
+                            if(typeof keyA === "undefined") {
+                                return 1;
+                            }
+                            const keyB = getSortableName(b);
+                            if(typeof keyB === "undefined") {
+                                return -1;
+                            }
+                            return keyA === keyB ? Comparison.EqualTo :
+                                typeof keyA === "string" && (typeof keyB ==="number" || typeof keyB ==="bigint") ? Comparison.LessThan :
+                                typeof keyB === "string" && (typeof keyA ==="number" || typeof keyA ==="bigint") ? Comparison.GreaterThan :
+                                keyA < keyB ? Comparison.LessThan :
+                                Comparison.GreaterThan;
+                        });
+                        clone.types = createNodeArray(types);
+                        visitNodes(clone.types, visitDeclarationSubtree);
+                        return cleanup(clone);
+                    }
                     default: Debug.assertNever(input, `Attempted to process unhandled node kind: ${(ts as any).SyntaxKind[(input as any).kind]}`);
                 }
             }
 
+            function getSortableName(n: TypeNode): string | number | bigint | undefined {
+                switch(n.kind){
+                    case SyntaxKind.TypePredicate:
+                    case SyntaxKind.TypeReference:
+                        const name = getTypeName((n as TypeReferenceNode).typeName);
+                        if (!name){
+                            return undefined;
+                        }
+                        return "000000_" + name;
+                    case SyntaxKind.LiteralType:
+                        const lt = n as LiteralTypeNode;
+                        return getLiteralTypeName(lt);
+                    case SyntaxKind.TupleType:
+                        const tt = n as TupleTypeNode;
+                        return "000001_" + tt.elementTypes.length + "_" + tt.elementTypes.map(getSortableName).join(", ");

```

Comment:
I like the idea of assigning each kind of type a rank and a string representation. I'm a bit worried about the efficiency of that approach, though, since the string representation will be built, used for comparison, then discarded. I think it's OK for an initial perf test, but I suspect it'll be more efficient to just recur through the type structure and early-exit when the correct Comparison can be determined.

For example, the sort function would take two parameters, and if `a.kind` is type predicate or type reference, and `b.kind` is anything else, then you can immediately `return -1`. If they're both type predicate or type reference, then you can compare their name. `getTypeName` can also be turned into `compareTypeName` so it doesn't have to construct strings.

This would save a lot of work when it comes to tuples, which currently need to recur, or numbers, which currently need to parse.